raiden-ts/src/transfers/epics/withdraw.ts

```diff
@@ -1,78 +1,370 @@
+ * @param deps.config$ - Config observable
+ * @returns Observable of withdraw.failure actions
+ */
+export const withdrawExpiresSoonEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { config$ }: RaidenEpicDeps,
+): Observable<withdraw.failure> =>
+  action$.pipe(
+    filter(withdrawMessage.request.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    withLatestFrom(config$),
+    mergeMap(([action, { revealTimeout }]) =>
+      action$.pipe(
+        filter(newBlock.is),
+        filter(({ payload }) => action.meta.expiration < payload.blockNumber + revealTimeout),
+        take(1),
+        mapTo(withdraw.failure(new RaidenError(ErrorCodes.WITHDRAW_EXPIRES_SOON), action.meta)),
+        // if withdrawMessage.success (WithdrawConfirmation) go through, don't fail
+        takeUntil(
+          action$.pipe(filter(isResponseOf<typeof withdrawMessage>(withdrawMessage, action.meta))),
+        ),
+      ),
+    ),
+  );
+
+/**
+ * Upon WithdrawConfirmation, send the on-chain withdraw request
+ *
+ * @param action$ - Observable of newBlock actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.address - Our address
+ * @param deps.signer - Signer instance
+ * @param deps.main - Main signer and address (if present)
+ * @param deps.log - Logger instance
+ * @param deps.getTokenNetworkContract - TokenNetwork contract getter
+ * @param deps.latest$ - Latest observable
+ * @returns Observable of withdrawExpired actions
+ */
+export const withdrawSendTxEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { address, signer, main, log, getTokenNetworkContract, latest$ }: RaidenEpicDeps,
+): Observable<withdraw.success | withdraw.failure> =>
+  action$.pipe(
+    filter(withdrawMessage.success.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    concatMap((action) =>
+      latest$.pipe(
+        first(),
+        mergeMap(({ state, config }) => {
+          const { subkey: configSubkey, revealTimeout } = config;
+          // don't send on-chain tx if we're 'revealTimeout' blocks from expiration
+          // this is our confidence threshold when we can get a tx inside timeout
+          assert(
+            action.meta.expiration >= state.blockNumber + revealTimeout,
+            ErrorCodes.WITHDRAW_EXPIRES_SOON,
+          );
+          const channel = state.channels[channelKey(action.meta)];
+          assert(channel?.state === ChannelState.open, 'channel not open');
+          assert(action.meta.totalWithdraw.gt(channel.own.withdraw), 'withdraw already performed');
+          const req = channel.own.pendingWithdraws.find(
+            (req) =>
+              req.type === MessageType.WITHDRAW_REQUEST &&
+              req.total_withdraw.eq(action.payload.message.total_withdraw) &&
+              req.expiration.eq(action.payload.message.expiration),
+          );
```

Comment:
I think I wrote already a comment to that above, not sure anymore. I know you don't like to have the epic into too small chunks of code all over the place. But things like this are common patterns you use all over the place. Exporting them to utility functions will reduce the complexity of the epics and reads much nicer without that many details. Duplicated code isn't helpful for such a huge code-base.